### PR TITLE
Adding tests for deconstruction

### PR DESCRIPTION
**Pull request by jcouv**
Some questions came up in LDM today:
- How do we resolve the `Deconstruct` method if one of the candidates is a generic `Deconstruct<T>`?
- The most straightforward deconstruction into locals with a `Deconstruct` method of the right types (no conversions) should not use temps (filed issue https://github.com/dotnet/roslyn/issues/13632 on that)
- A simple deconstruction of a tuple literal could be optimized to not construct a `ValueTuple`, but use temps instead (filed issue https://github.com/dotnet/roslyn/issues/13631 on that)

FYI @jaredpar 
@dotnet/roslyn-compiler for review.

### CODE REVIEW

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. Only tests in this PR.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 1]
```diff
@@ -398,6 +455,29 @@ public void Deconstruct(out int a, out string b)
...
+        c = new[] { 2, 2 };
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source);
+            comp.VerifyDiagnostics(
+                // (4,58): error CS1108: A parameter cannot have all the specified modifiers; there are too many modifiers on the parameter
+                //     public void Deconstruct(out int a, out string b, out params int[] c)
+                Diagnostic(ErrorCode.ERR_MultiParamMod, "params").WithLocation(4, 58)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: This error is not specific to `Deconstruct` though, is it?

Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Correct. 
If "out params" were supported, then we'd want to figure out the proper rules for resolving Deconstruct. But "out params" isn't supported :-)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 2]
```diff
@@ -1355,6 +1435,71 @@ public static void Deconstruct(this C value, out int a, out string b)
...
+        a = 2;
+        b = default(T);
+    }
+}";
+
+            var comp = CreateCompilationWithMscorlib(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef, SystemCoreRef });
+            comp.VerifyDiagnostics(
+                // (8,18): error CS0411: The type arguments for method 'Extension.Deconstruct<T>(C, out int, out T)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //         (x, y) = new C();
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "new C()").WithArguments("Extension.Deconstruct<T>(C, out int, out T)").WithLocation(8, 18),
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Why don't we infer `string`?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 3]
```diff
@@ -1355,6 +1435,71 @@ public static void Deconstruct(this C value, out int a, out string b)
...
+        public void AmbiguousDeconstructGenericExtensionMethod()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        long x;
+        string y;
+        (x, y) = new C();
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Consider adding a call to the other overload so it's clear both can be used, `(x, x) = new C();`, and verify execution.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs (hunk 2 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: We look-up Deconstruct as-if you invoked `Deconstruct(out var x1, out var x2, ...)`. Any types we may know from the left-hand-side don't contribute.
We discussed it in LDM earlier this week (which is why I wanted to record the tests) and decided that is ok for now.

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs (hunk 3 above)
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Resolved in person. I'll update the name of the test.

**Pull request merged**